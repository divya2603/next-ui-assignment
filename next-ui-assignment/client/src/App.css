.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App {
  min-height: 100vh;
  background-color: #282c34;
}
.App-header {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

a {
  color: inherit;
  text-decoration: auto;
  display: inline-block;
}
.results-container {
  width: 100%;
  height: 50vh;
  border: 2px solid;
  overflow: scroll;
}
.list-item {
  display: flex;
  justify-content: center;
  margin: 10px;
  cursor: pointer;
}
.list-item .item-name{
  margin: 0px 5px;
}
.item-link {
  padding: 5px 10px;
  border: 1px solid;
  border-radius: 20px;
  font-size: 16px;
}
.meta-info {
  font-size: 8px;
  text-align: right;
  margin-right: 10px;
  margin-top: 2px;
}
.Artist {
  color:aquamarine;
}
.Album {
  color: darkseagreen;
}
.Song {
  color:salmon;
}