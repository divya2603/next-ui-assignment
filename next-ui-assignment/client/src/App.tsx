import React, { useState, useEffect, useRef, useMemo, lazy, Suspense } from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from "redux-thunk";
import { AnyAction } from 'redux';
import styled from 'styled-components';

import { getData, getNextPageData, clearData } from './actions/apiCall';
import { debounce } from "./utils";
import {
  ExtraDispatchArguments,
  Props,
  StateProps,
  EventProp
} from './interfaces'
import {
  HEADER_NAME
} from './Constants';
import './App.css';

const SearchResults = lazy(() => import('./Components/SearchForm'));

const Wrapper = styled.section`
padding: 4em;
`;
const Input = styled.input`
padding: 0.5em;
margin: 0.5em;
width: 100%;
border: none;
border-radius: 3px;
`;
const Form = styled.form`
width: 50%;
margin: 0 auto;
`;
const Header = styled.header``;

const mapStateToProps = (state: StateProps) => {
  return {
    searchData: state.searchData.data
  }
}

const mapDispatchToProps = (dispatch: ThunkDispatch<StateProps, ExtraDispatchArguments, AnyAction>)  => {
  return {
    getData: ({ query }) => dispatch(getData({ query })),
    getNextPageData:  ({ query, pageNumber }) => dispatch(getNextPageData({ query, pageNumber })),
    clearData: () => dispatch(clearData())
  }
}

function App(props: Props): JSX.Element {
  const { searchData, getData, getNextPageData, clearData } = props;
  const { data, pageNumber } = searchData || {};

  const [query, setQuery] = useState<string>('');
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [pageNum, setPageNum] = useState<number>(0);

  const containerRef = useRef(null);

  const options = {
    root: null,
    rootMargin: "0px",
    threshold: 1
  }

  useEffect(() => {
    const observer = new IntersectionObserver(scrollObserver, options);
    if(containerRef.current) {
      observer.observe(containerRef.current);
    }
    return () => {
      if(containerRef.current) {
        observer.unobserve(containerRef.current);
      }
    }
  },[containerRef, options]);

  const scrollObserver = (entries) => {
    const [entry] = entries;
    setIsVisible(entry.isIntersecting);
  }

  useEffect(() => {
    setPageNum(pageNumber);
  },[isVisible]);

  useEffect(() => {
    query.length && getNextPageData({ query, pageNumber });
  }, [pageNum]);

  useEffect(() => {
    if (query.length > 0) {
      getData({query});
    } else {
      clearData();
    }
  },[query]);

  const handleSearch = () => {
    console.log(" query ", query);   
  }

  const handleChange = (event: EventProp): void => {
    setQuery(event.target.value);
  }

  const debouncedChangeHandler = useMemo(() => {
    return debounce(handleChange, 300);
  }, []);

  return (
    <Wrapper className="App">
      <Header className="App-header"> {HEADER_NAME} </Header>
      <Form 
        className="App-header" 
        id="search-form" 
        onSubmit={e => {
        e.preventDefault();
        handleSearch();
      }}>
        <Input 
          type="text"
          onChange={debouncedChangeHandler} 
          placeholder="Search artists, albums, and/or songs... "
        />
        <Suspense fallback={<div>Loadingâ€¦</div>}>
          <SearchResults 
            data={data} 
            containerRef={containerRef} 
            query={query}
          />
        </Suspense>
      </Form>
    </Wrapper>
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
