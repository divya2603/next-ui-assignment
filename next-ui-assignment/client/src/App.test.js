import React from 'react';
import ReactDOM from 'react-dom';
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import App from './App';
import searchData from './testData';
import styled from 'styled-components';

const Input = styled.input;

const mockStore = configureMockStore({});
configure({adapter: new Adapter()});

const mockFunc = {
  handleSearch: jest.fn(),
  preventDefault: jest.fn()
};

const getPage = renderType => {
  const store = mockStore({
      searchData: searchData
  });

  const Page = (
      <Provider store={store}>
          <App />
      </Provider>
  );
  if (renderType === 'mount') {
      return mount(Page);
  }
  return shallow(Page);
};

describe('Main Component', () => {
  let page;
  it('should render with given state from Redux store', () => {
    page = getPage();
    expect(page).toMatchSnapshot();
  });
});