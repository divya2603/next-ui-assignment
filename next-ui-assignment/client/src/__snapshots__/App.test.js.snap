// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component should render with given state from Redux store 1`] = `ShallowWrapper {}`;

exports[`Reducers should render with given state from Redux store 1`] = `ShallowWrapper {}`;
